twitter_api/main.py

```diff
@@ -17,12 +21,214 @@ def before_request():
 
 # implement your views here
 
+@json_only
+@app.route('/login', methods = ['POST'])
+def login():    # jon working on
+    # Inputs: username and password
+    # Ouputs: generated new access token for given username.
+    authdata = request.get_json()
+    
+    try:
+        username = authdata['username']
+        password = authdata['password']
+        password = password.encode('utf-8')
+        password = md5(password).hexdigest()
+    except KeyError:
+        abort(400)
 
+    db_authdata = g.db.execute('SELECT password, id from user WHERE username = ?', (username,))
+    pass_and_id = db_authdata.fetchone()
+    if pass_and_id:
+        passdata, uid = pass_and_id
+    else:
+        passdata = None
+        
+    if not passdata: # no entry in DB.  should be 400, but tests asks for 404.
+        abort(404)
+    
+    if passdata != password:
+        abort(401)
+    
+    usertoken = generate_token()
+    
+    g.db.execute('INSERT INTO auth (user_id, access_token) VALUES (?, ?)', (uid, usertoken))
+    g.db.commit()
+    
+    return Response(json.dumps(dict(access_token = usertoken)), status=201, content_type=JSON_MIME_TYPE)
+
```

Comment:
Flask has a built-in function called "jsonify" which builds a fitting response object for you. 

Nice code overall.